subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/GradlePluginDevelopmentExtension.java

```diff
@@ -62,7 +62,7 @@ public GradlePluginDevelopmentExtension(Project project, SourceSet pluginSourceS
     }
 
     public GradlePluginDevelopmentExtension(Project project, SourceSet pluginSourceSet, SourceSet[] testSourceSets) {
-        this.plugins = project.container(PluginDeclaration.class);
+        this.plugins = project.getObjects().container(PluginDeclaration.class);
         this.pluginSourceSet = pluginSourceSet;
```

Comment:
let's change the places that instantiate some of the container types directly, to use the factory methods (either use the internal or object factory for these).

- `DefaultConfigurationPublications`
- `DefaultPluginManager`
- `DefaultConfiguration`
- `DefaultVariant`
- `MainExecutableVariant` (should use the public factory, as this is in a plugin).
- `MainLibraryVariant` (should use the public factory)
- `BaseBinarySpec`
- `DefaultPrebuiltLibrary`
- `CachingLibraryBinaryLocator`
- `ProjectLibraryBinaryLocator`
- `DefaultClassDirectoryBinarySpec`
- `Sign` (should use the public factory)

Reply:
Done.